rfc/1216-resolve-cloud-tokens.md

```diff
@@ -0,0 +1,845 @@
+# Resolve Cloud Tokens
+
+* **Original Author(s):** @iliapolo
+* **Tracking Issue**: https://github.com/cdk8s-team/cdk8s/issues/1216
+* **API Bar Raiser**: @rix0rrr
+
+Users can now configure custom resolvers to control how cdk8s resolves values before writing 
+them to the manifest. In addition, Custom resolvers for the AWS CDK and CDKTF token system are available, 
+allowing users to author cdk8s applications that rely on cloud resources.
+
+## Working Backwards
+
+### CHANGELOG
+
+`feat(synth): custom resolvers`
+
+### README
+
+Custom resolvers are a mechanism to inject custom logic into the cdk8s value resolution process. 
+It allows to transform any value just before being written to the Kubernetes manifest.
+
+To define a custom resolver, first create a class that implements the `ITokenResolver` interface:
+
+```ts
+import { ITokenResolver, ResolutionContext } from 'cdk8s';
+
+export class MyCustomResolver implements ITokenResolver {
+
+  public resolve(context: ResolutionContext, value: any): any {
+    // run some custom logic
+  }
+
+}
+```
+
+The `context` argument contains information about the value that is currently being resolved:
+
+- **obj**: `ApiObject` currently being resolved.
+- **key**: Array containing the JSON path elements of they keys leading up to the value.
+
+When you create a cdk8s `Chart`, pass the resolver instance to it via the `resolver` property:
+
+```ts
+import { App, Chart } from 'cdk8s'
+
+const app = new App();
+new Chart(app, 'Chart', { resolver: new MyCustomResolver() });
+```
+
+When you run `cdk8s synth`, your custom logic will be invoked, and the return 
+value will replace the original value and be written to the manifest. For example, if 
+you define a Kubernetes service like so:
+
+```ts
+new KubeService(this, 'Service', {
+  spec: {
+    type: 'LoadBalancer',
+  }
+});
+```
+
+Your resolver will be invoked with the following arguments:
+
+- **context**
+  - *obj*: The `KubeService` instance of type `ApiObject`.
+  - *key*: `['spec', 'type']`
+- **value**: `LoadBalancer`
+
+
+One common use-case for this feature is to automatically resolve deploy time 
+attributes of cloud resources, and pass them to your Kubernetes workloads. To that end, 
+cdk8s provides a resolver for the following CDK frameworks:
+
+#### AWS Cloud Development Kit
+
+In this example, we create an S3 `Bucket` with the AWS CDK, and pass its (deploy time generated) name 
+as an environment variable to a Kubernetes `CronJob` resource.
+
+```ts
+import * as aws from 'aws-cdk-lib';
+import * as k8s from 'cdk8s';
+import * as kplus from 'cdk8s-plus-26';
+
+import { AwsCdkTokenResolver } from '@cdk8s/aws-cdk-token-resolver';
+
+const awsApp = new aws.App();
+const stack = new aws.Stack(awsApp, 'Stack');
+
+const resolver = new AwsCdkTokenResolver(stack);
+
+const k8sApp = new k8s.App();
```

Comment:
Wasn't the original vision to have everything in the same `App` ?

Reply:
Not really. As we talked offline, the only single `App` that makes sense is a sort of "constructs" App that can synthesize two different artifacts, which we don't have. But even that doesn't really fit this use case because conceptually this is about passing information from one app to another, as they each have a separate synthesis and deployment lifecycle.